### PR TITLE
Tackle many warnings

### LINKED ISSUE #3871: BUG: `Cofunction += Cofunction` leads to `DeprecationWarning`s
**Describe the bug**
```
import warnings

warnings.simplefilter("default")

import firedrake as fd

mesh = fd.UnitIntervalMesh(1)
space = fd.FunctionSpace(mesh, "DG", 0)

u = fd.Cofunction(space.dual())
u += fd.Cofunction(space.dual())
```
leads to the warnings
```
/home/jmaddis2/build/firedrake/firedrake/src/PyOP2/pyop2/types/dat.py:408: DeprecationWarning: Expression.index(i) is deprecated and will be removed in 2H2025. Use 'expr[i] 'instead.
  lhs = _self.index(i)
/home/jmaddis2/build/firedrake/firedrake/src/PyOP2/pyop2/types/dat.py:416: DeprecationWarning: Expression.index(i) is deprecated and will be removed in 2H2025. Use 'expr[i] 'instead.
  rhs = _other.index(i)
```

**Expected behavior**
No warning.

**Environment:**
Ubuntu 24.04, new Firedrake build.

### PR DESCRIPTION
**Pull request by connorjward**
## TODOs
- [x] Use actual upstream loopy
- [x] Check that this isn't a major performance regression (!?) (might just be overloaded runners)

Closes #3871

(Isn't it nice that I don't need a PyOP2 PR!)

### CODE REVIEW

File: .github/workflows/pyop2.yml
Comment (connorjward) [inline .github/workflows/pyop2.yml:111]:
```suggestion
```

File: pyop2/codegen/rep2loopy.py
Comment (JHopeCollins) [inline pyop2/codegen/rep2loopy.py:222]: Why `object.__setattr__` and not `self.parameters = ...` or `setattr`?

File: pyop2/codegen/rep2loopy.py
Comment (JHopeCollins) [inline pyop2/codegen/rep2loopy.py:297]: This ensures ordering but also changes the return type.
Why not:
```python
return {i: None for i in indices}.keys()
```
Or
```python
return type(indices)(dict.fromkeys(indices))
```

File: pyop2/codegen/rep2loopy.py
Comment (JHopeCollins) [inline pyop2/codegen/rep2loopy.py:418]: Personally I find `dict.update` clearer than `|=`.

File: pyop2/codegen/rep2loopy.py
Comment (JHopeCollins) [inline pyop2/codegen/rep2loopy.py:554]: Please open a Firedrake issue linking to these lines and the loopy issue to we can keep track and hopefully revery once loopy gets fixed.

File: pyop2/codegen/rep2loopy.py
Comment (connorjward) [reply pyop2/codegen/rep2loopy.py:222]: Because this is a frozen dataclass. I was copying what loopy do internally ([example](https://github.com/inducer/loopy/blob/76f4afc27f343bc7e95a11b9b04d30c574402bcd/loopy/symbolic.py#L646), I really don't like it).

Comment (JHopeCollins) [reply pyop2/codegen/rep2loopy.py:222]: Does it break otherwise? Is there something specific about frozen dataclasses that means you have to do this?
I would usually advocate for using the smallest hammer possible!

Comment (connorjward) [reply pyop2/codegen/rep2loopy.py:222]: Yeah otherwise it breaks unfortunately

Comment (JHopeCollins) [reply pyop2/codegen/rep2loopy.py:222]: `setattr` breaks or `self.parameters = `? What's the error?

Comment (connorjward) [reply pyop2/codegen/rep2loopy.py:222]: My guess is that dataclasses overload `__setattr__` and so we need to specify `object.__setattr__`. Will report back.

File: pyop2/codegen/rep2loopy.py
Comment (connorjward) [reply pyop2/codegen/rep2loopy.py:297]: I need to return a type that can act as an ordered set. The return type needs to be changed because I continue to need ordered set semantics outside of the function.

File: pyop2/codegen/rep2loopy.py
Comment (connorjward) [reply pyop2/codegen/rep2loopy.py:554]: Won't be long! https://github.com/inducer/loopy/pull/895
I might just wait for that to be merged.

File: pyop2/codegen/rep2loopy.py
Comment (JHopeCollins) [reply pyop2/codegen/rep2loopy.py:297]: Ok, that's fine then.

File: pyop2/codegen/rep2loopy.py
Comment (inducer) [reply pyop2/codegen/rep2loopy.py:222]: That's exactly right. There's no *real* immutability in Python, and so the MO of frozen dataclasses is to make mutation stick out like a sore thumb. Reading your comments above, I would say mission accomplished. :slightly_smiling_face: 

The conventional, easier-on-the-eyes approach would be to just delete the constructor (and let the `@dataclass` transform worry about writing they eyesore code). This is not feasible here because the argument order would mismatch what dataclasses generates, breaking compatibility. :cry:

Comment (JHopeCollins) [reply pyop2/codegen/rep2loopy.py:222]: I see. Not the nicest thing but doesn't sound like there's much alternative if this is a dataclass but the `PyOP2KernelCallable` has a different signature to `PyOP2KernelCallable`.

File: pyop2/codegen/rep2loopy.py
Comment (connorjward) [reply pyop2/codegen/rep2loopy.py:554]: Fixed upstream. Will remove this.

**Review by JHopeCollins — APPROVED** [review]
LGTM, once the workflow branches are removed.
Thanks @connorjward!

**Pull request merged**
